src/eredis_cluster.erl

```diff
@@ -175,63 +175,91 @@ query(Command, PoolKey) ->
     Transaction = fun(Worker) -> qw(Worker, Command) end,
     query(Transaction, Slot, 0).
 
-query(_, _, ?REDIS_CLUSTER_REQUEST_TTL) ->
+query(_, _, ?QUERY_LIMIT) ->
     {error, no_connection};
 query(Transaction, Slot, Counter) ->
-    %% Throttle retries
-    throttle_retries(Counter),
+    exponential_backoff(Counter),
 
-    {Pool, Version} = eredis_cluster_monitor:get_pool_by_slot(Slot),
+    Pool = eredis_cluster_monitor:get_pool_by_slot(Slot),
+    Result = try_query(Pool, Transaction),
 
-    Result = eredis_cluster_pool:transaction(Pool, Transaction),
-    case handle_transaction_result(Result, Version) of 
+    case handle_transaction_result(Result) of
         retry -> query(Transaction, Slot, Counter + 1);
-        Result -> Result
+        Payload -> Payload
     end.
 
-handle_transaction_result(Result, Version) ->
-    case Result of 
-       % If we detect a node went down, we should probably refresh the slot
-        % mapping.
-        {error, no_connection} ->
-            eredis_cluster_monitor:refresh_mapping(Version),
-            retry;
+try_query(Pool, Transaction) ->
+    Result = eredis_cluster_pool:transaction(Pool, Transaction),
+
+    case Result of
+        % When a redirection is encountered, it is likely multiple slots
+        % were reconfigured rather than just one, so updating the client
+        % configuration as soon as possible is often the best strategy
+        % https://redis.io/topics/cluster-spec
+        {error, <<"MOVED ", SlotAddressPort/binary>>} ->
+            eredis_cluster_monitor:async_refresh_mapping(),
+            try_redirect(SlotAddressPort, Transaction);
+
+        % ASK is like MOVED but the difference is that the client should retry
+        % against on new instance only this query, not next queries. That means:
+        % smart clients should not update internal state
+        % https://redis.io/presentation/Redis_Cluster.pdf
```

Comment:
I think we can give a better explanation to why this is necessary.
We do not call `async_refresh_mapping/0` here because even though the client has already detected that information is being `MOVED` from node `A` to `B` and is already asking `B`, the information has not been copied from `A` to `B` yet. Therefore Redis is telling us to `ASK A` only this time, given that the next time we fetch this information, it will probably be in `B` already.